arangod/Cluster/ClusterMethods.cpp

```diff
@@ -2754,4 +2756,813 @@ int fetchEdgesFromEngines(std::string const& dbname,
+  }  
+
+  return arangodb::Result();
+
+}
+
+arangodb::Result lockDBServerTransactions(
+  std::string const& backupId, std::vector<ServerID> const& dbServers,
+  std::chrono::steady_clock::time_point const& end) {
+
+  using namespace std::chrono;
+
+  VPackBuilder lock;
+  {
+    VPackObjectBuilder o(&lock);
+    lock.add("id", VPackValue(backupId));
+  }
+  auto body = std::make_shared<std::string const>(lock.toJson());
+  auto cc = ClusterComm::instance();
+  double lockWait = 1.0;
+
+  std::unordered_map<std::string, std::string> headers;
+
+  while (cc != nullptr && end < steady_clock::now()) {
+
+    CoordTransactionID coordTransactionID = TRI_NewTickServer();
+    std::string const lockUrl =
+      std::string("/_admin/hotbackup/lock?timeout=") + std::to_string(lockWait);
+
+    auto lockWaitEnd = steady_clock::now() + seconds((uint64_t)lockWait);
+
+    std::unordered_map<std::string, OperationID> lockRes;
+    for (auto const& dbServer : dbServers) {
+      cc->asyncRequest(
+        "", coordTransactionID, "server:" + dbServer, RequestType::POST, lockUrl, body,
+        headers, nullptr, lockWait);
+    }
+
+    // not stopping && not timed out
+    uint64_t nLocks;
+    while (cc != nullptr && lockWaitEnd < steady_clock::now()) {
+
+      // Collect all call backs one after the other
+      auto const waitRes = cc->wait("", coordTransactionID, 0, "",
+                                    duration<double>(lockWaitEnd - steady_clock::now()).count());
+
+      if (waitRes.status == CL_COMM_SENT) {
+        auto wrBody = waitRes.result->getBodyVelocyPack();
+        VPackSlice slc = wrBody->slice();
+        // Sanity check
+        if (slc.isObject()
+            && slc.hasKey("id") && slc.get("id").isString()
+            && slc.hasKey("locked")   && slc.get("locked").isBoolean()) {
+          std::string otherBackupId = slc.get("id").copyString();
+          bool locked = (slc.get("locked").getBool());
+          if (backupId != otherBackupId) {
+            std::stringstream s;
+            s << std::string("Lock result contains wrong backup ID. Mine:");
+            s << backupId + " " + waitRes.serverID + "'s " + otherBackupId + ". Failing!";
+            LOG_TOPIC(ERR, Logger::HOTBACKUP) << s.str();
+            return arangodb::Result(TRI_ERROR_INTERNAL, s.str());
+          }
+          if (!locked) {
+            std::stringstream s;
+            s <<"Lock attempt rejected by ";
+            s << waitRes.serverID + ". Failing!";
+            LOG_TOPIC(ERR, Logger::HOTBACKUP) << s.str();
+            return arangodb::Result(TRI_ERROR_INTERNAL, s.str());
+          }
+        }
+        if (++nLocks == dbServers.size()) {
+          break;
+        }
+      } else {
+        std::stringstream s;
+        s << "Failed to send lock request ";
+        s << backupId + " to " + waitRes.serverID; 
+        LOG_TOPIC(ERR, Logger::HOTBACKUP) << s.str();
+        return arangodb::Result(TRI_ERROR_INTERNAL, s.str());
+      }
+    }
+
```

Comment:
Do we not have to cancel the locks on all servers, where we have succeeded, if any of the locks were not achieved until the timeout happens? Is using the timeout the right method here? Or should we rather have the other side return within a prescribed time, regardless on whether the lock has succeeded or not? This logic seems to be too simple.

Reply:
Done